zcash_client_backend/src/data_api/chain.rs

```diff
@@ -236,46 +239,49 @@ where
 /// # Ok(())
 /// # }
 /// ```
-pub fn scan_cached_blocks<E, N, P, C, D>(
+pub fn scan_cached_blocks<'a, E, N, P, C, D>(
     params: &P,
     cache: &C,
-    data: &mut D,
+    data: Arc<Mutex<D>>,
     limit: Option<u32>,
-) -> Result<(), E>
+) -> Box<dyn Future<Item = (), Error = E> + Send + 'a>
 where
-    P: consensus::Parameters,
+    P: consensus::Parameters + Send + Sync,
     C: BlockSource<Error = E>,
     D: WalletWrite<Error = E, NoteRef = N>,
-    N: Copy + Debug,
-    E: From<Error<N>>,
+    N: Copy + Debug + Send,
+    E: From<Error<N>> + Send + 'a,
 {
-    let sapling_activation_height = params
+    let data_clone = data.lock().unwrap();
```

Comment:
`data_clone` is not a very good name for this, you can call it `data_guard`

Reply:
done